packages/ipfs-core/src/components/add-all/index.js

```diff
@@ -13,10 +13,10 @@ const mergeOptions = require('merge-options').bind({ ignoreUndefined: true })
  * @param {import('..').GCLock} config.gcLock
  * @param {import('..').Preload} config.preload
  * @param {import('..').Pin} config.pin
- * @param {import('../init').ConstructorOptions<any, boolean>} config.options
+ * @param {ShardingOptions} [config.options]
  */
-module.exports = ({ block, gcLock, preload, pin, options: constructorOptions }) => {
-  const isShardingEnabled = constructorOptions.EXPERIMENTAL && constructorOptions.EXPERIMENTAL.sharding
+module.exports = ({ block, gcLock, preload, pin, options }) => {
+  const isShardingEnabled = options && options.sharding
```

Comment:
Working on this pull request actually shifted my opinion towards disliking optionals (for non end-user APIs), because in refactor if you accidentally omit an option(s) type checker tells you nothing code still type checks even though behavior changes. So then it’s up to tests to catch the problem. More specifically I’m finding that it’s better to define following type:

```js
type Encoding = {
  format?: string
}
```

As 

```js
type Encoding {
  format: undefined | string
}
```

That means a bit more typing for the caller but in return mistake of omitting `format` is caught and reported by type checker. I recall I did something along those lines where I introduced regressions that luckly were caught by tests.

Reply:
yes non user apis should always have defaults defined at the edge between user input and internal logic. i would even say remove the undefined and make it required after that edge.